[FIX/#174] Group / 포토피커 적용

TaewoongR opened the pull request:
- closed #174

## *📍 Work Description*

- 포토피커 적용
- firestore 함수 수정
- 백핸들러 설정
- 예외별 토스트 메세지 처리
- 이미지 규격 설정

## *📸 Screenshot*


https://github.com/user-attachments/assets/e02563db-b72f-4f1b-945a-e0032b5f2d9d


## *📢 To Reviewers*
- 사인님 커스텀 포토피커를 그룹 생성화면과 그룹 편집 화면에 적용했습니다.
- firestore에 그룹 collection에 저장하기 전에 firebase storage에 이미지를 저장/편집을 한 후에 이미지 저장 주소를 포함하여 그룹 collection에 저장합니다.
- 백핸들러를 통해 디바이스 뒤로가기 버튼을 누를 때 그룹 생성/편집 화면으로 돌아옵니다.
- 이미지, 제목, 설명 조건을 충족시키지 않고 생성을 하면 토스트 메세지로 경고를 띄웁니다.
- 모든 사진 사이즈는 coil의 contentSalce 속성에 ContentScale.FillBounds 를 대입하여 모든 사진 영역이 경계에 꽉 차서 전부 보이도록 했습니다.

## ⏲️Time

    - 5시간

TaewoongR committed:
fix/#174 : 그룹 생성 포토피커 적용

TaewoongR committed:
fix/#174 : 그룹 생성 firestore 요청 함수 수정

- firebase storage 사진 저장 적용
- 사진 url을 포함하여 firestore group collection 에 저장

TaewoongR committed:
fix/#174 : 그룹 편집 포토피커 적용

TaewoongR committed:
fix/#174 : 그룹 편집 firestore 요청 함수 수정

- 기존 사진 url과 firestore의 사진 url 대조 후 사진이 편집되었으면 firebase의 storage에 덮어씌우기
- 변경된 storage 사진 url을 포함하여 firestore group collection 에 저장

TaewoongR committed:
fix/#174 : for ktlint

TaewoongR committed:
feat/#174 : back handler 적용

- 포토피커는 네비게이션에 종속되지 않았으므로 백핸들러로 디바이스 뒤로가기 터치시 생성/편집 화면으로 돌아가기

TaewoongR committed:
fix/#174 : 그룹 생성 토스트 메세지 설정

- 필요없는 state 속성 삭제

TaewoongR committed:
fix/#174 : coil 이미지 contentScale 조정

- 규격 bound에 모든 사진 영역이 가득 차도록 설정

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupCreationScreen.kt:
```diff
@@ -46,50 +50,86 @@ import com.boostcamp.mapisode.mygroup.intent.GroupCreationIntent
+		viewModel.onIntent(GroupCreationIntent.Initialize)
+	}
+
+	BackHandler {
+		if (uiState.value.isSelectingGroupImage) {
+			viewModel.onIntent(GroupCreationIntent.OnBackToGroupCreation)
+		} else {
+			viewModel.onIntent(GroupCreationIntent.OnBackClick)
+		}
+	}
```
조금 더 안전한 동작을 위해 도입했다고 보는 게 맞을까요?

TaewoongR replied:
다른 팀원분들이 적용하고 계셔서 저도 따라 적용했습니다.
포토피커가 네비게이션에 종속되지 않아서 디바이스 뒤로가기를 눌렀을 때 Route 기준으로 이전화면으로 가는걸 방지합니다.

Ameri-Kano commented on data/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/GroupRepositoryImpl.kt:
```diff
@@ -189,10 +203,41 @@ class GroupRepositoryImpl @Inject constructor(private val database: FirebaseFire
+			val existingGroup = groupDocRef.get().await().toObject(GroupFirestoreModel::class.java)
+				?: throw Exception()
+
+			// 앱에서 설정한 이미지와 firebase storage URL 비교하여 다르면 업로드
+			val isImageUrlDifferent = existingGroup.imageUrl != groupModel.imageUrl
+
+			val updatedImageUrl = if (isImageUrlDifferent) {
+				val imageRef = storage.reference.child("group/${groupModel.id}/1")
+				val uploadTask = imageRef.putFile(groupModel.imageUrl.toUri()).await()
+				uploadTask.task.result.storage.downloadUrl.await().toString()
+			} else {
+				existingGroup.imageUrl
+			}
```
이 부분이 조금 궁금한데, 변경하지 않는다면 모델의 imageUrl이 기존에 업로드한 이미지의 url을 가지게 되나요?

TaewoongR replied:
네 맞아요!
최종적으로 서버에 올라가는 이미지 url 은 storage 의 url 이에요.
coil 에서는 로컬, 서버 url 둘다 처리를 하고 있어서 model의 url은 두가지 경우중 하나입니다.

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupCreationScreen.kt:
```diff
@@ -107,7 +147,7 @@ fun GroupCreationContent(
 		topBar = {
 			TopAppBar(
-				title = "그룹 편집",
+				title = "그룹 생성",
```
별 건 아니지만 리소스 처리!

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupCreationScreen.kt:
```diff
@@ -166,14 +216,17 @@ fun GroupCreationField(
-						onClick = { },
-						showImage = false,
+						onClick = { onPhotoPickerClick() },
+						showImage = profileUrl == "",
```
```suggestion
						showImage = profileUrl.isEmpty(),
```
제공하는 메서드를 사용하는 것도 좋을 것 같아요!

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupCreationScreen.kt:
```diff
@@ -247,7 +300,7 @@ fun GroupCreationField(
 					onGroupEditClick(name, description, profileUrl)
 				},
-				text = "편집하기",
+				text = "생성하기",
```
여기도!

Ameri-Kano commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/viewmodel/GroupCreationViewModel.kt:
```diff
@@ -34,32 +42,75 @@ class GroupCreationViewModel @Inject constructor(
+				copy(
+					isInitializing = false,
+					group = group.copy(
+						id = UUID.randomUUID().toString().replace("-", ""),
```
이건 저만 생각한 개인적인 계획인데, core:common 모듈에 이 함수를 만들어 둬야 할 것 같아요! 
(하면 제가 할겁니다)

Ameri-Kano commented on gradle/libs.versions.toml:
```diff
@@ -79,6 +79,7 @@ junit = "4.13.2"
 androidx-junit = "1.2.1"
 espresso-core = "3.6.1"
 material = "1.12.0"
+firebaseStorage = "21.0.1"
```
firebase 번들 쓰면 될 것 같은데 따로 추가해야 하는 필요가 있었나요?

TaewoongR replied:
어허...IDE quick action 쓰지 말아야했는데 멋대로 추가해줬네요. 이거 제거할게요

Ameri-Kano reviewed (approved):
수고 많으셨습니다!

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupCreationScreen.kt:
```diff
@@ -46,50 +50,86 @@ import com.boostcamp.mapisode.mygroup.intent.GroupCreationIntent
 	viewModel: GroupCreationViewModel = hiltViewModel(),
 ) {
+	val context = LocalContext.current
 	val uiState = viewModel.uiState.collectAsStateWithLifecycle()
 	val effect = rememberFlowWithLifecycle(
 		flow = viewModel.sideEffect,
 		initialValue = GroupCreationSideEffect.Idle,
 	).value
 
+	if (uiState.value.isInitializing) {
+		viewModel.onIntent(GroupCreationIntent.Initialize)
+	}
```
``` kotlin
	val uiState by viewModel.uiState.collectAsStateWithLifecycle()
```

로 하시면 직접 value에 접근하지 않아도 됩니다.

TaewoongR replied:
수정 진행해볼게요

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/screen/GroupEditScreen.kt:
```diff
@@ -73,29 +87,56 @@ fun GroupEditScreen(
+				Toast.makeText(context, effect.messageResId, Toast.LENGTH_SHORT).show()
+			}
+
+			else -> {}
```
처리하고 있지 않은 SideEffect가 있나요?

TaewoongR replied:
제가 side Effect 를 Screen에서 stateFlow로 바꾸고 collectAsStateWithLifecycle 와 동일한 처리를 해서 초기값을 Idle로 주고 있어요.
굳이 stateFlow로 바꿀 필요가 없는데 번거롭게 했네요.

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/state/GroupCreationState.kt:
```diff
@@ -1,7 +1,12 @@
 @Immutable
-data class GroupCreationState(val isGroupEditError: Boolean = false) : UiState
+data class GroupCreationState(
+	val isInitializing: Boolean = true,
```
저도 이렇게 한 부분이 있는데, false로 두고 초기화 호출시 true로 변경해야 좋을 것 같아요

TaewoongR replied:
피드백 적용해볼게요

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/viewmodel/GroupCreationViewModel.kt:
```diff
@@ -34,32 +42,75 @@ class GroupCreationViewModel @Inject constructor(
+							),
+						)
+					}
+					Timber.e("Group: ${currentState.group}")
```
로그는 제거해주세요

haeti-dev commented on feature/mygroup/src/main/java/com/boostcamp/mapisode/mygroup/viewmodel/GroupCreationViewModel.kt:
```diff
@@ -69,7 +120,9 @@ class GroupCreationViewModel @Inject constructor(
 				}
 			} catch (e: Exception) {
-				intent { copy(isGroupEditError = true) }
+				postSideEffect(GroupCreationSideEffect.ShowToast(R.string.message_error_edit_group))
+				delay(100)
+				postSideEffect(GroupCreationSideEffect.Idle)
```
delay를 준 이유가 있으신지요?

TaewoongR replied:
side effect를 state flow로 변환한 까닭에 상태를 한번더 변환시켜줘야 연속으로 동일한 오류가 뜰 때 수집가능하도록 하고 있어요.
delay는 수집되고 난 후에 처리 중간에 다음 상태가 수집되어서 취소되는걸 방지해주는 역할이에요.
결과적으로 state flow로 바꾸면 안되었다...ㅋㅋㅋ

haeti-dev reviewed (approved):
고생하셨습니다~

TaewoongR committed:
fix/#174 : 코드리뷰 피드백 적용

- IDE 자동 gradle 추가 제거
- 문자열 strings 리소스 처리
- uiState 선언부에 by 적용하여 value에 곧바로 접근
- initialize 실행 시점을 상태가 아닌 이벤트로
